File: app/src/main/java/zmuzik/slidingpuzzle2/gamescreen/GameActivity.java
package zmuzik.slidingpuzzle2.gamescreen;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.graphics.Bitmap;
import android.graphics.Point;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.Display;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;
import android.widget.ProgressBar;

import com.squareup.picasso.MemoryPolicy;
import com.squareup.picasso.NetworkPolicy;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import zmuzik.slidingpuzzle2.App;
import zmuzik.slidingpuzzle2.R;
import zmuzik.slidingpuzzle2.common.Keys;
import zmuzik.slidingpuzzle2.common.PreferencesHelper;
import zmuzik.slidingpuzzle2.common.ShakeDetector;
import zmuzik.slidingpuzzle2.common.Toaster;
import zmuzik.slidingpuzzle2.flickr.FlickrApi;

public class GameActivity extends Activity implements GameScreenView, ShakeDetector.OnShakeListener {

    final String TAG = this.getClass().getSimpleName();

    @BindView(R.id.board)
    PuzzleBoardView mBoard;
    @BindView(R.id.progressBar)
    ProgressBar mProgressBar;
    @BindView(R.id.shuffleBtn)
    ImageView mShuffleBtn;

    int mScreenWidth;
    int mScreenHeight;
    int mBoardWidth;
    int mBoardHeight;

    @Inject
    PreferencesHelper mPrefsHelper;
    @Inject
    Toaster mToaster;
    @Inject
    FlickrApi mFlickrApi;
    @Inject
    GameScreenPresenter mPresenter;
    @Inject
    ShakeDetector mShakeDetector;

    private GameActivityComponent mComponent;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        inject();
        setContentView(R.layout.activity_game);
        ButterKnife.bind(this);
        Intent intent = getIntent();
        Animation shake = AnimationUtils.loadAnimation(this, R.anim.shake_anim);
        mShuffleBtn.startAnimation(shake);
        if (intent == null) {
            mToaster.show(R.string.picture_not_supplied);
            finish();
        }

        setScreenOrientation(getIntent().getExtras().getBoolean(Keys.IS_HORIZONTAL));
        resolveScreenDimensions();

        mProgressBar.setVisibility(View.VISIBLE);
        //mBoard.setVisibility(View.GONE);
        mPresenter.requestPictureUri(intent);

    }

    @Override
    protected void onResume() {
        super.onResume();
        if (mShakeDetector != null) {
            mShakeDetector.register();
            mShakeDetector.setOnShakeListener(this);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mShakeDetector != null) {
            mShakeDetector.unRegister();
        }
    }

    @Override
    public void onShake() {
        shuffle(null);
    }

    public void finishWithMessage(int stringId) {
        mToaster.show(stringId);
        finish();
    }

    public void loadPicture(String uri) {
        Picasso.with(this)
                .load(uri)
                .memoryPolicy(MemoryPolicy.NO_STORE)
                .networkPolicy(NetworkPolicy.NO_STORE)
                .resize(mScreenWidth, mScreenHeight)
                .centerInside()
                .into(mTarget);
    }

    void inject() {
        mComponent = DaggerGameActivityComponent.builder()
                .appComponent(((App) getApplication()).getComponent(this))
                .gameActivityModule(new GameActivityModule(this))
                .build();
        mComponent.inject(this);
        mComponent.inject(mPresenter);
    }

    public GameActivityComponent getComponent() {
        return mComponent;
    }

    void setScreenOrientation(boolean isHorizontal) {
        setRequestedOrientation(isHorizontal
                ? ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
                : ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
    }

    void resolveScreenDimensions() {
        Display display = getWindowManager().getDefaultDisplay();
        Point size = new Point();
        display.getSize(size);
        mScreenWidth = size.x;
        mScreenHeight = size.y;
    }

    public int getMaxScreenDim() {
        return (mScreenWidth > mScreenHeight) ? mScreenWidth : mScreenHeight;
    }

    public void shuffle(View v) {
        mShuffleBtn.setVisibility(View.GONE);
        if (mBoard != null) mBoard.maybeShuffle();
    }

    void adjustBoardDimensions(PuzzleBoardView board, Bitmap bitmap) {
        float screenSideRatio = (float) mScreenWidth / mScreenHeight;
        float origPictureSideRatio = (float) bitmap.getWidth() / bitmap.getHeight();
        if (origPictureSideRatio > screenSideRatio) {
            mBoardWidth = mScreenWidth;
            mBoardHeight = (int) (mScreenWidth / origPictureSideRatio);
        } else {
            mBoardHeight = mScreenHeight;
            mBoardWidth = (int) (mScreenHeight * origPictureSideRatio);
        }
        int widthMultiple = (mBoardWidth > mBoardHeight)
                ? mPrefsHelper.getGridDimLong()
                : mPrefsHelper.getGridDimShort();
        int heightMultiple = (mBoardWidth > mBoardHeight)
                ? mPrefsHelper.getGridDimShort()
                : mPrefsHelper.getGridDimLong();
        mBoardWidth = mBoardWidth - (mBoardWidth % widthMultiple);
        mBoardHeight = mBoardHeight - (mBoardHeight % heightMultiple);

        board.setDimensions(mBoardWidth, mBoardHeight);
    }

    @Override
    public void onStop() {
        Picasso.with(this).cancelRequest(mTarget);
        super.onStop();
    }

    Target mTarget = new Target() {
        @Override
        public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
            adjustBoardDimensions(mBoard, bitmap);
            mBoard.setBitmap(bitmap);
            mShuffleBtn.setVisibility(View.VISIBLE);
            mProgressBar.setVisibility(View.GONE);
            mBoard.setVisibility(View.VISIBLE);
        }

        @Override
        public void onBitmapFailed(Drawable errorDrawable) {
            mProgressBar.setVisibility(View.GONE);
            mToaster.show(R.string.unable_to_load_flickr_picture);
            finish();
        }

        @Override
        public void onPrepareLoad(Drawable placeHolderDrawable) {
        }
    };
}

File: app/src/main/java/zmuzik/slidingpuzzle2/gamescreen/GameActivity.kt
package zmuzik.slidingpuzzle2.gamescreen

import android.app.Activity
import android.content.pm.ActivityInfo
import android.graphics.Bitmap
import android.graphics.Point
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.View
import android.view.animation.AnimationUtils
import butterknife.ButterKnife
import com.squareup.picasso.MemoryPolicy
import com.squareup.picasso.NetworkPolicy
import com.squareup.picasso.Picasso
import com.squareup.picasso.Target
import kotlinx.android.synthetic.main.activity_game.*
import zmuzik.slidingpuzzle2.App
import zmuzik.slidingpuzzle2.R
import zmuzik.slidingpuzzle2.common.Keys
import zmuzik.slidingpuzzle2.common.PreferencesHelper
import zmuzik.slidingpuzzle2.common.ShakeDetector
import zmuzik.slidingpuzzle2.common.Toaster
import zmuzik.slidingpuzzle2.flickr.FlickrApi
import javax.inject.Inject

class GameActivity : Activity(), GameScreenView, ShakeDetector.OnShakeListener {

    internal val TAG = this.javaClass.simpleName

    internal var screenWidth: Int = 0
    internal var screenHeight: Int = 0
    internal var boardWidth: Int = 0
    internal var boardHeight: Int = 0

    @Inject
    lateinit var prefs: PreferencesHelper
    @Inject
    lateinit var toaster: Toaster
    @Inject
    lateinit var flickrApi: FlickrApi
    @Inject
    lateinit var presenter: GameScreenPresenter
    @Inject
    lateinit var shakeDetector: ShakeDetector

    lateinit var component: GameActivityComponent

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        inject()
        setContentView(R.layout.activity_game)
        ButterKnife.bind(this)
        if (intent == null) {
            toaster.show(R.string.picture_not_supplied)
            finish()
        }

        setScreenOrientation(intent.extras.getBoolean(Keys.IS_HORIZONTAL))
        resolveScreenDimensions()

        progressBar.visibility = View.VISIBLE
        intent?.let { presenter.requestPictureUri(it) }
    }

    override fun onResume() {
        super.onResume()
        shakeDetector.register()
        shakeDetector.setOnShakeListener(this)
    }

    override fun onPause() {
        super.onPause()
        shakeDetector.unRegister()
    }

    override fun onShake() {
        shuffle(null)
    }

    override fun finishWithMessage(stringId: Int) {
        toaster.show(stringId)
        finish()
    }

    override fun loadPicture(uri: String) {
        Picasso.with(this)
                .load(uri)
                .memoryPolicy(MemoryPolicy.NO_STORE)
                .networkPolicy(NetworkPolicy.NO_STORE)
                .resize(screenWidth, screenHeight)
                .centerInside()
                .into(mTarget)
    }

    internal fun inject() {
        component = DaggerGameActivityComponent.builder()
                .appComponent((application as App).getComponent(this))
                .gameActivityModule(GameActivityModule(this))
                .build()
        component.inject(this)
        component.inject(presenter)
    }

    internal fun setScreenOrientation(isHorizontal: Boolean) {
        requestedOrientation = if (isHorizontal)
            ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
        else
            ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
    }

    internal fun resolveScreenDimensions() {
        val display = windowManager.defaultDisplay
        val size = Point()
        display.getSize(size)
        screenWidth = size.x
        screenHeight = size.y
    }

    override fun getMaxScreenDim() = if (screenWidth > screenHeight) screenWidth else screenHeight

    override fun hideShuffleIcon() {
        shuffleBtn.visibility = View.GONE
    }

    fun shuffle(v: View?) {
        hideShuffleIcon()
        board?.maybeShuffle()
    }

    internal fun adjustBoardDimensions(board: PuzzleBoardView, bitmap: Bitmap) {
        val screenSideRatio = screenWidth.toFloat() / screenHeight
        val origPictureSideRatio = bitmap.width.toFloat() / bitmap.height
        if (origPictureSideRatio > screenSideRatio) {
            boardWidth = screenWidth
            boardHeight = (screenWidth / origPictureSideRatio).toInt()
        } else {
            boardHeight = screenHeight
            boardWidth = (screenHeight * origPictureSideRatio).toInt()
        }
        val widthMultiple = if (boardWidth > boardHeight) prefs.gridDimLong else prefs.gridDimShort
        val heightMultiple = if (boardWidth > boardHeight) prefs.gridDimShort else prefs.gridDimLong

        boardWidth -= boardWidth % widthMultiple
        boardHeight -= boardHeight % heightMultiple

        board.setDimensions(boardWidth, boardHeight)
    }

    public override fun onStop() {
        Picasso.with(this).cancelRequest(mTarget)
        super.onStop()
    }

    internal var mTarget: Target = object : Target {
        override fun onBitmapLoaded(bitmap: Bitmap, from: Picasso.LoadedFrom) {
            adjustBoardDimensions(board, bitmap)
            board.setBitmap(bitmap)
            progressBar.visibility = View.GONE
            board.visibility = View.VISIBLE
            val shake = AnimationUtils.loadAnimation(this@GameActivity, R.anim.shake_anim)
            shuffleBtn.visibility = View.VISIBLE
            shuffleBtn.startAnimation(shake)
        }

        override fun onBitmapFailed(errorDrawable: Drawable?) {
            progressBar.visibility = View.GONE
            toaster.show(R.string.unable_to_load_flickr_picture)
            finish()
        }

        override fun onPrepareLoad(placeHolderDrawable: Drawable?) {}
    }
}
